provisioner/terraform/tfparse/tfparse.go

```diff
@@ -0,0 +1,490 @@
+	}
+
+	// iterate through module.DataResources to get the default values for all
+	// coder_parameter data resources.
+	paramsDefaults = make(map[string]string)
+	for _, dataResource := range module.DataResources {
+		if dataResource.Type != "coder_parameter" {
+			logger.Debug(ctx, "skip resource as it is not a coder_parameter", "resource_name", dataResource.Name, "resource_type", dataResource.Type)
+		}
+
+		var file *hcl.File
+		var diags hcl.Diagnostics
+		parser := hclparse.NewParser()
+
+		if !strings.HasSuffix(dataResource.Pos.Filename, ".tf") {
+			logger.Debug(ctx, "only .tf files can be parsed", "filename", dataResource.Pos.Filename)
+			continue
+		}
+
+		// We know in which HCL file is the data resource defined.
+		file, diags = parser.ParseHCLFile(dataResource.Pos.Filename)
+		if diags.HasErrors() {
+			return nil, nil, xerrors.Errorf("can't parse the resource file %q: %s", dataResource.Pos.Filename, diags.Error())
+		}
+
+		// Parse root to find "coder_parameter".
+		content, _, diags := file.Body.PartialContent(rootTemplateSchema)
+		if diags.HasErrors() {
+			return nil, nil, xerrors.Errorf("can't parse the resource file: %s", diags.Error())
+		}
+
+		// Iterate over blocks to locate the exact "coder_parameter" data resource.
+		for _, block := range content.Blocks {
+			if !slices.Equal(block.Labels, []string{"coder_parameter", dataResource.Name}) {
+				continue
+			}
+
+			// Parse "coder_parameter" to find the default value.
+			resContent, _, diags := block.Body.PartialContent(coderParameterSchema)
+			if diags.HasErrors() {
+				return nil, nil, xerrors.Errorf(`can't parse the coder_parameter: %s`, diags.Error())
+			}
+
+			if _, ok := resContent.Attributes["default"]; !ok {
+				paramsDefaults[dataResource.Name] = ""
+			} else {
+				expr := resContent.Attributes["default"].Expr
+				value, err := previewFileContent(expr.Range())
+				if err != nil {
+					return nil, nil, xerrors.Errorf("can't preview the resource file: %v", err)
+				}
+				paramsDefaults[dataResource.Name] = strings.Trim(value, `"`)
+			}
+		}
+	}
+	return varsDefaults, paramsDefaults, nil
+}
+
+// EvalProvisionerTags evaluates the given workspaceTags based on the given
+// default values for variables and coder_parameter data sources.
+func EvalProvisionerTags(varsDefaults, paramsDefaults, workspaceTags map[string]string) (map[string]string, error) {
+	// Filter only allowed data sources for preflight check.
+	// This is not strictly required but provides a friendlier error.
+	if err := validWorkspaceTagValues(workspaceTags); err != nil {
+		return nil, err
+	}
+	// We only add variables and coder_parameter data sources. Anything else will be
+	// undefined and will raise a Terraform error.
+	evalCtx := buildEvalContext(varsDefaults, paramsDefaults)
+	tags := make(map[string]string)
+	for workspaceTagKey, workspaceTagValue := range workspaceTags {
+		expr, diags := hclsyntax.ParseExpression([]byte(workspaceTagValue), "expression.hcl", hcl.InitialPos)
```

Comment:
Is the idea to replace `hclsyntax.ParseExpression` in wsbuilder?

Reply:
Yes, ideally. I'd like to have all of this logic in one place.